File: backend/services/media/src/main/java/app/mnema/media/config/SecurityConfig.java
package app.mnema.media.config;

import app.mnema.media.security.InternalTokenAuthFilter;
import app.mnema.media.security.MediaInternalAuthProps;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpMethod;
import org.springframework.security.oauth2.server.resource.web.BearerTokenResolver;
import org.springframework.http.HttpHeaders;
import org.springframework.security.config.Customizer;
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.oauth2.server.resource.web.BearerTokenAuthenticationFilter;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;

import java.util.List;

@Configuration
@EnableMethodSecurity
@EnableConfigurationProperties({CorsProps.class, MediaInternalAuthProps.class})
public class SecurityConfig {

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http,
                                                   CorsConfigurationSource corsConfigurationSource,
                                                   InternalTokenAuthFilter internalTokenAuthFilter,
                                                   BearerTokenResolver bearerTokenResolver) throws Exception {
        http
                .cors(cors -> cors.configurationSource(corsConfigurationSource))
                .csrf(AbstractHttpConfigurer::disable)
                .authorizeHttpRequests(auth -> auth
                        .requestMatchers("/actuator/health/**", "/actuator/info", "/actuator/prometheus").permitAll()
                        .requestMatchers("/error").permitAll()
                        .requestMatchers(HttpMethod.POST, "/resolve").permitAll()
                        .requestMatchers(HttpMethod.OPTIONS, "/**").permitAll()
                        .anyRequest().authenticated()
                )
                .oauth2ResourceServer(oauth2 -> oauth2
                        .jwt(Customizer.withDefaults())
                        .bearerTokenResolver(bearerTokenResolver)
                );

        http.addFilterBefore(internalTokenAuthFilter, BearerTokenAuthenticationFilter.class);

        return http.build();
    }

    @Bean
    public CorsConfigurationSource corsConfigurationSource(CorsProps props) {
        CorsConfiguration cfg = new CorsConfiguration();
        var origins = (props.origins() == null || props.origins().isEmpty())
                ? List.of("https://mnema.app", "http://localhost:3005")
                : props.origins();

        cfg.setAllowedOrigins(origins);
        cfg.setAllowedMethods(List.of("GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"));
        cfg.setAllowedHeaders(List.of("*"));
        cfg.setExposedHeaders(List.of("Authorization", "Content-Type"));
        cfg.setAllowCredentials(true);
        cfg.setMaxAge(3600L);

        UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
        source.registerCorsConfiguration("/**", cfg);
        return source;
    }

    @Bean
    public BearerTokenResolver bearerTokenResolver(MediaInternalAuthProps props) {
        return request -> {
            String authHeader = request.getHeader(HttpHeaders.AUTHORIZATION);
            if (authHeader == null || !authHeader.startsWith("Bearer ")) {
                return null;
            }
            String token = authHeader.substring("Bearer ".length());
            if (token.equals(props.internalToken())) {
                return null;
            }
            return token;
        };
    }

    @Bean
    public InternalTokenAuthFilter internalTokenAuthFilter(MediaInternalAuthProps props) {
        return new InternalTokenAuthFilter(props);
    }
}

File: backend/services/user/src/main/kotlin/app/mnema/user/config/SecurityConfig.kt
package app.mnema.user.config

import app.mnema.user.security.InternalTokenAuthFilter
import app.mnema.user.security.UserInternalAuthProps
import org.springframework.boot.actuate.autoconfigure.security.servlet.EndpointRequest
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.env.Environment
import org.springframework.http.HttpHeaders
import org.springframework.security.config.Customizer
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.oauth2.server.resource.web.BearerTokenResolver
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.access.intercept.AuthorizationFilter

@Configuration
@EnableMethodSecurity
@EnableConfigurationProperties(UserInternalAuthProps::class)
class SecurityConfig(
    private val env: Environment
) {
    @Bean
    fun securityFilterChain(
        http: HttpSecurity,
        internalTokenAuthFilter: InternalTokenAuthFilter,
        bearerTokenResolver: BearerTokenResolver
    ): SecurityFilterChain {
        val swaggerEnabled = env.getProperty("springdoc.swagger-ui.enabled", Boolean::class.java, true)
        val apiDocsEnabled = env.getProperty("springdoc.api-docs.enabled", Boolean::class.java, true)
        val publicMatchers = mutableListOf("/actuator/**")
        if (swaggerEnabled || apiDocsEnabled) {
            publicMatchers.addAll(
                listOf(
                    "/v3/api-docs/**",
                    "/swagger-ui/**",
                    "/swagger-ui.html"
                )
            )
        }

        http
            .cors {  }
            .csrf { it.disable() }
            .authorizeHttpRequests {
                if (publicMatchers.isNotEmpty()) {
                    it.requestMatchers(*publicMatchers.toTypedArray()).permitAll()
                }
                it.requestMatchers(EndpointRequest.to("health", "info")).permitAll()
                it.anyRequest().authenticated()
            }
            .oauth2ResourceServer {
                it.jwt(Customizer.withDefaults())
                    .bearerTokenResolver(bearerTokenResolver)
            }

        http.addFilterBefore(internalTokenAuthFilter, AuthorizationFilter::class.java)
        return http.build()
    }

    @Bean
    fun bearerTokenResolver(props: UserInternalAuthProps): BearerTokenResolver = BearerTokenResolver { request ->
        val authHeader = request.getHeader(HttpHeaders.AUTHORIZATION)
        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
            return@BearerTokenResolver null
        }
        val token = authHeader.removePrefix("Bearer ").trim()
        if (token == props.internalToken) {
            return@BearerTokenResolver null
        }
        token
    }

    @Bean
    fun internalTokenAuthFilter(props: UserInternalAuthProps): InternalTokenAuthFilter =
        InternalTokenAuthFilter(props)
}
